alvr/server/cpp/platform/linux/EncodePipeline.cpp

```diff
@@ -111,17 +58,24 @@ alvr::EncodePipeline::~EncodePipeline()
   avcodec_free_context(&encoder_ctx);
 }
 
-bool alvr::EncodePipeline::GetEncoded(std::vector<uint8_t> &out, uint64_t *pts)
+bool alvr::EncodePipeline::GetEncoded(FramePacket &packet)
 {
-  AVPacket * enc_pkt = av_packet_alloc();
-  int err = avcodec_receive_packet(encoder_ctx, enc_pkt);
-  if (err == AVERROR(EAGAIN)) {
-    return false;
-  } else if (err) {
+  encoder_packet = av_packet_alloc();
+  int err = avcodec_receive_packet(encoder_ctx, encoder_packet);
+  if (err != 0) {
+    av_packet_free(&encoder_packet);
+    if (err == AVERROR(EAGAIN)) {
+      return false;
+    }
     throw alvr::AvException("failed to encode", err);
   }
-  filter_NAL(enc_pkt->data, enc_pkt->size, out);
-  *pts = enc_pkt->pts;
-  av_packet_free(&enc_pkt);
+  packet.data = encoder_packet->data;
+  packet.size = encoder_packet->size;
+  packet.pts = encoder_packet->pts;
   return true;
 }
+
+void alvr::EncodePipeline::Free()
```

Comment:
I would say instead of adding `EncodePipeline::Free()` just free the packet in `GetEncoded()` if needed.

Reply:
But how can I do this when `GetEncoded()` only gets the packet and checks whether it's available? Basically, we get the packet, do some processing, then deallocate. I cannot get the packet and deallocate immediately, that would invalidate the pointer.

Reply:
```cpp
bool alvr::EncodePipeline::GetEncoded(FramePacket &packet)
{
  av_packet_free(&encoder_packet);
  encoder_packet = av_packet_alloc();
  // ...
```

Reply:
Makes sense, nice.